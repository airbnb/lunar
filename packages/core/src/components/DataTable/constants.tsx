import { WidthProperties } from './types';

export const SELECTION_OPTIONS = {
  ACTIVE: 'ACTIVE',
  DISABLED: 'DISABLED',
  INACTIVE: 'INACTIVE',
  HAS_ACTIVE_CHILD: 'HAS_ACTIVE_CHILD',
};

export const STATUS_OPTIONS = {
  ALERT: 'ALERT',
  WARNING: 'WARNING',
};

type HeightMap = {
  [key: string]: number;
};

export const HEIGHT_TO_PX: HeightMap = {
  micro: 32,
  small: 48,
  regular: 56,
  large: 64,
  jumbo: 108,
};

export const DEFAULT_WIDTH_PROPERTIES: WidthProperties = {
  width: 50,
  flexGrow: 1,
  flexShrink: 1,
  minWidth: 0,
};

export const EXPANDABLE_COLUMN_WIDTH = 50;

export const SELECTABLE_COLUMN_WIDTH = 50;
export const SELECTABLE_COLUMN_WIDTH_EXPANDABLE = 39;
