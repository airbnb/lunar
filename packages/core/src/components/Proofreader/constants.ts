import { ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT } from '../../keys';

export const AIRBNB_REGEX = /\b(((air|ari|iar)[bn]{3})(?!\.com))\b/gi;
export const NON_WORD_REGEX = /\W/;

export const ARROW_KEYS = [ARROW_LEFT, ARROW_UP, ARROW_DOWN, ARROW_RIGHT];

export const NO_LOCALE = 'none';
export const AUTO_DETECT_LOCALE = 'auto';
export const LOCALE_TO_LT_LOCALE: { [locale: string]: string } = {
  de: 'de-DE',
  en: 'en-US',
  ja: 'ja-JP',
  pt: 'pt-PT',
  ru: 'ru-RU',
  zh: 'zh-CN',
};
