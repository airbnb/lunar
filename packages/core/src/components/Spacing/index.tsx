import React from 'react';
import { StyleBlock } from 'aesthetic';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type SpacingRange =
  | 0
  | 0.5
  | 1
  | 1.5
  | 2
  | 2.5
  | 3
  | 3.5
  | 4
  | 4.5
  | 5
  | 5.5
  | 6
  | 6.5
  | 7
  | 7.5
  | 8
  | 8.5
  | 9
  | 9.5
  | 10
  | 10.5
  | 11
  | 11.5
  | 12;

export type Props = {
  /** Apply spacing on all sides. */
  all?: SpacingRange;
  /** Apply spacing on the bottom. */
  bottom?: SpacingRange;
  /** Content to render. */
  children: NonNullable<React.ReactNode>;
  /** Apply spacing on the left and right. */
  horizontal?: SpacingRange;
  /** Render as inline block instead of block. */
  inline?: boolean;
  /** Use padding instead of margin. */
  inner?: boolean;
  /** Apply spacing on the left. */
  left?: SpacingRange;
  /** Apply spacing on the right. */
  right?: SpacingRange;
  /** Base element / component type. Can be a string representing a DOM element. */
  tag?: 'article' | 'div' | 'footer' | 'header' | 'section';
  /** Apply spacing on the top. */
  top?: SpacingRange;
  /** Apply spacing on the top and bottom. */
  vertical?: SpacingRange;
};

function cleanClassName(className: string | number) {
  return String(className).replace('.', 'dot');
}

/** Common component for arbitray layout structure and spacing. */
export class Spacing extends React.Component<Props & WithStylesProps> {
  static defaultProps: Partial<Props> = {
    all: 0,
    bottom: 0,
    horizontal: 0,
    inline: false,
    inner: false,
    left: 0,
    right: 0,
    top: 0,
    vertical: 0,
  };

  render() {
    const {
      cx,
      all,
      bottom,
      children,
      horizontal,
      left,
      tag: TagProp = 'div',
      top,
      inline,
      inner,
      right,
      styles,
      vertical,
    } = this.props;
    const type = inner ? 'inner' : 'outer';
    const classes: any = [];

    if (all) {
      const cleaned = cleanClassName(all);
      classes.push(
        styles[`${type}Top_${cleaned}`],
        styles[`${type}Right_${cleaned}`],
        styles[`${type}Bottom_${cleaned}`],
        styles[`${type}Left_${cleaned}`],
      );
    } else {
      if (vertical) {
        const cleaned = cleanClassName(vertical);
        classes.push(styles[`${type}Top_${cleaned}`], styles[`${type}Bottom_${cleaned}`]);
      } else {
        if (top) {
          const cleaned = cleanClassName(top);
          classes.push(styles[`${type}Top_${cleaned}`]);
        }

        if (bottom) {
          const cleaned = cleanClassName(bottom);
          classes.push(styles[`${type}Bottom_${cleaned}`]);
        }
      }

      if (horizontal) {
        const cleaned = cleanClassName(horizontal);
        classes.push(styles[`${type}Left_${cleaned}`], styles[`${type}Right_${cleaned}`]);
      } else {
        if (left) {
          const cleaned = cleanClassName(left);
          classes.push(styles[`${type}Left_${cleaned}`]);
        }

        if (right) {
          const cleaned = cleanClassName(right);
          classes.push(styles[`${type}Right_${cleaned}`]);
        }
      }
    }

    return (
      <TagProp className={cx(...classes, inline && styles.spacing_inline)}>{children}</TagProp>
    );
  }
}

export default withStyles(({ unit }) => {
  const spacing: { [key: string]: StyleBlock } = {};

  for (let i = 0; i <= 12; i += 0.5) {
    const size = unit * i;

    ['Top', 'Right', 'Bottom', 'Left'].forEach(side => {
      const cleaned = cleanClassName(i);
      spacing[`outer${side}_${cleaned}`] = { [`margin${side}`]: size };
      spacing[`inner${side}_${cleaned}`] = { [`padding${side}`]: size };
    });
  }

  return {
    ...spacing,

    spacing_inline: {
      display: 'inline-block',
    },
  };
})(Spacing);
