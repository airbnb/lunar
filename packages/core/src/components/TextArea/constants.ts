// eslint-disable-next-line import/prefer-default-export
export const LANGUAGE_RULES = {
  UNNECESSARY_ADMIN_NOTE_RULE: 'AGENT_TOOLS_UNNECESSARY_ADMIN_NOTE_RULE',
};
