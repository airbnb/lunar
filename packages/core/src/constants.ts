import { Currency, Locale, LocaleDefinition, TimeZone } from './types';

// I18N/L10N
export const DEFAULT_LOCALE: Locale = 'en';
export const DEFAULT_CURRENCY: Currency = 'USD';
export const DEFAULT_TIMEZONE: TimeZone = 'UTC';

export const LOCALES: Locale[] = [
  'de', // German
  'en', // English
  'es', // Spanish
  'fr', // French
  'it', // Italian
  'ja', // Japanese
  'ko', // Korean
  'ms', // Malay
  'nl', // Dutch
  'pt', // Portuguese
  'ru', // Russian
  'tr', // Turkish
  'zh', // Mandarin
];

// Used for Google Translate API
export const TRANSLATE_LOCALES: LocaleDefinition[] = [
  { locale: 'ar', label: 'ÿπÿ±ÿ®Ÿâ', name: 'Arabic' },
  { locale: 'ca', label: 'Catal√†', name: 'Catalan' },
  { locale: 'cs', label: 'ƒåe≈°tina', name: 'Czech' },
  { locale: 'da', label: 'Dansk', name: 'Danish' },
  { locale: 'de', label: 'Deutsch', name: 'German' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en', label: 'English', name: 'English' },
  { locale: 'es', label: 'Espa√±ol', name: 'Spanish' },
  { locale: 'es-419', label: 'Espa√±ol (Am√©rica Latina)', name: 'Spanish (Latin America)' },
  { locale: 'fi', label: 'Suomalainen', name: 'Finnish' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'he', label: '◊¢÷¥◊ë◊®÷¥◊ô◊™', name: 'Hebrew' },
  { locale: 'hr', label: 'Hrvatski', name: 'Croatian' },
  { locale: 'hu', label: 'Magyar', name: 'Hungarian' },
  { locale: 'id', label: 'Bahasa Indonesia', name: 'Indonesian' },
  { locale: 'is', label: '√çslensku', name: 'Icelandic' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  { locale: 'ko', label: 'ÌïúÍµ≠Ïñ¥', name: 'Korean' },
  { locale: 'ms', label: 'Melayu', name: 'Malay' },
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'no', label: 'Norsk', name: 'Norwegian' },
  { locale: 'pl', label: 'Polskie', name: 'Polish' },
  { locale: 'pt', label: 'Portugu√™s', name: 'Portuguese' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'ru', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  { locale: 'th', label: '‡πÑ‡∏ó‡∏¢', name: 'Thai' },
  { locale: 'tr', label: 'T√ºrk√ße', name: 'Turkish' },
  { locale: 'vi', label: 'Ti·∫øng Vi·ªát', name: 'Vietnamese' },
  { locale: 'zh', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
  { locale: 'zh-TW', label: '‰∏≠Êñá (ÁπÅ)', name: 'Chinese (Traditional)' },
];

// Used for Language Tool (not all are listed)
// https://languagetool.org/http-api/swagger-ui/#!/default/get_languages
export const LT_LOCALES: LocaleDefinition[] = [
  { locale: 'ca-ES', label: 'Catal√†', name: 'Catalan' },
  { locale: 'ca-ES-valencia', label: 'Catal√† (Valenci√†)', name: 'Catalan' },
  { locale: 'da-DK', label: 'Dansk', name: 'Danish' },
  { locale: 'de-AT', label: 'Deutsch (√ñsterreich)', name: 'German (Austria)' },
  { locale: 'de-DE', label: 'Deutsch (Deutschland)', name: 'German (Germany)' },
  { locale: 'el', label: 'ŒïŒªŒªŒ∑ŒΩŒπŒ∫Œ¨', name: 'Greek' },
  { locale: 'en-AU', label: 'English (Australia)' },
  { locale: 'en-CA', label: 'English (Canada)' },
  { locale: 'en-GB', label: 'English (Great Britain)' },
  { locale: 'en-NZ', label: 'English (New Zealand)' },
  { locale: 'en-US', label: 'English (United States)' },
  { locale: 'es', label: 'Espa√±ol (Espa√±a)', name: 'Spanish (Spain)' },
  { locale: 'fr', label: 'Fran√ßais', name: 'French' },
  { locale: 'it', label: 'Italiano', name: 'Italian' },
  { locale: 'ja-JP', label: 'Êó•Êú¨Ë™û', name: 'Japanese' },
  // Korean ÌïúÍµ≠Ïñ¥
  // Malay Melayu
  { locale: 'nl', label: 'Nederlands', name: 'Dutch' },
  { locale: 'pl-PL', label: 'Polskie', name: 'Polish' },
  { locale: 'pt-PT', label: 'Portugu√™s (Portugal)', name: 'Portuguese (Portugal)' },
  { locale: 'pt-BR', label: 'Portugu√™s (Brasil)', name: 'Portuguese (Brazil)' },
  { locale: 'ro-RO', label: 'Rom√¢nƒÉ', name: 'Romanian' },
  { locale: 'ru-RU', label: '–†—É—Å—Å–∫–∏–π', name: 'Russian' },
  { locale: 'sv', label: 'Svenska', name: 'Swedish' },
  // Turkish T√ºrk√ße
  { locale: 'uk-UA', label: '–£–∫—Ä–∞—ó–Ω—Å—å–∫–∞', name: 'Ukrainian' },
  { locale: 'zh-CN', label: '‰∏≠Êñá (ÁÆÄ)', name: 'Chinese (Simplified)' },
];

// DATE/TIME
export const FORMAT_YMD = 'yyyy-MM-dd';
export const FORMAT_PREFIX_DAY = 'ccc'; // Sun
export const FORMAT_SUFFIX_TIMEZONE = 'ZZZZ'; // PST

// MISC
export const MICROS = 1000000;

// THEMES
export const STATUSES = ['notice', 'info', 'success', 'warning', 'danger', 'muted'];
export const BRANDS = ['luxury', 'plus'];
export const Z_INDEX_MODAL = 2000;
export const Z_INDEX_PORTAL = 2000; // Same as Modal to fix Tooltip in Modal
export const Z_INDEX_TOAST = 3000;

// EMOJIS
export const EMOJI_WHITELIST = [
  '1F600', // üòÄ grinning face
  '1F60A', // üòä smiling face with smiling eyes
  '1F642', // üôÇ slightly smiling face
  '1F389', // üéâ party popper
  '2728', // ‚ú® sparkles
  '1F334', // üå¥ palm tree
  '1F333', // üå≥ deciduous tree
  '1F335', // üåµ cactus
  '2600', // ‚òÄÔ∏è sun
  '2744', // ‚ùÑÔ∏è snowflake
  '1F3E0', // üè†Ô∏è house
  '1F697', // üöó automobile
  '2708', // ‚úàÔ∏è airplane
  '1F682', // üöÇ locomotive
  '26F5', // ‚õµÔ∏è sailboat
];

// REACT
export const PASSTHROUGH_WRAPPER_NAMES = ['Tooltip', 'Permission', 'Trebuchet', 'ERF', 'Memo'];
export const STRIP_HOC_NAMES = ['Memo', 'Proxy', 'withStyles', 'withTheme', 'connect'];
