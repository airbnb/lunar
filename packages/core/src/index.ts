import Aesthetic, { FontFace } from 'aesthetic';
import AphroditeAesthetic, { NativeBlock, ParsedBlock } from 'aesthetic-adapter-aphrodite';
import { Settings as LuxonSettings } from 'luxon';
import { Path as EmojiPath } from 'interweave-emoji';
import globalStyles from './themes/global';
import lightTheme from './themes/light';
import darkTheme from './themes/dark';
import getLocaleFromClient from './utils/getLocaleFromClient';
import getTimezoneFromClient from './utils/getTimezoneFromClient';
import { DEFAULT_LOCALE, DEFAULT_TIMEZONE } from './constants';
import {
  Locale,
  TimeZone,
  Logger,
  Translator,
  TranslateParams,
  TranslateProps,
  Theme,
} from './types';

export type Settings = {
  defaultLocale?: Locale;
  defaultTimezone?: TimeZone;
  emojiCDN?: EmojiPath;
  errorURL?: string;
  fontFaces?: { [fontFamily: string]: FontFace[] };
  fontFamily?: string;
  logger?: Logger | null;
  name: string;
  theme?: 'light' | 'dark';
  translator?: Translator | null;
  translatorComponent?: React.ComponentType<TranslateProps> | null;
};

class Core {
  settings: Required<Settings> = {
    defaultLocale: getLocaleFromClient() || DEFAULT_LOCALE,
    defaultTimezone: getTimezoneFromClient() || DEFAULT_TIMEZONE,
    emojiCDN: '',
    errorURL: '',
    fontFaces: {},
    fontFamily:
      '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    logger: null,
    name: 'Lunar',
    theme: 'light',
    translator: null,
    translatorComponent: null,
  };

  protected aesthetic?: Aesthetic<Theme, NativeBlock, ParsedBlock>;

  initialize(settings: Settings) {
    this.settings = {
      ...this.settings,
      ...settings,
    };

    this.bootstrapAesthetic();
    this.bootstrapLuxon();
  }

  bootstrapAesthetic() {
    const { theme, fontFaces } = this.settings;
    const fontFamily = this.fontFamily();
    const globals = globalStyles(fontFaces);

    this.aesthetic = new AphroditeAesthetic<Theme>([], {
      theme,
      passThemeProp: false,
      pure: true,
    })
      .registerTheme('light', lightTheme(fontFamily), globals)
      .registerTheme('dark', darkTheme(fontFamily), globals);
  }

  bootstrapLuxon() {
    LuxonSettings.defaultLocale = this.locale();
    LuxonSettings.defaultZoneName = this.timezone();
    LuxonSettings.throwOnInvalid = true;
  }

  getAesthetic() {
    if (__DEV__) {
      if (!this.aesthetic) {
        throw new Error(
          'Aesthetic has not been initialized. Please call `Core.initialize()` from `@airbnb/lunar`.',
        );
      }
    }

    return this.aesthetic!;
  }

  fontFamily(): string {
    const locale = this.locale();

    if (locale.startsWith('ja')) {
      return '"ヒラギノ角ゴ Pro", "Hiragino Kaku Gothic Pro", メイリオ, Meiryo, Osaka, "ＭＳ Ｐゴシック", "MS PGothic", "MS Gothic", "ＭＳ ゴシック", "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.startsWith('ko')) {
      return '"나눔 고딕", "Nanum Gothic", "맑은 고딕", "Malgun Gothic", "Apple Gothic", 돋움, Dotum, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    if (locale.includes('zh')) {
      return '"Hiragino Sans GB", 华文细黑, STHeiti, 微软雅黑, "Microsoft YaHei", SimHei, "Helvetica Neue", Helvetica, Arial, sans-serif';
    }

    return this.settings.fontFamily;
  }

  locale(): Locale {
    return this.settings.defaultLocale || DEFAULT_LOCALE;
  }

  log = (error: Error, extra: object) => {
    const { logger } = this.settings;

    if (logger) {
      logger(error, extra);
    }
  };

  timezone(): TimeZone {
    return this.settings.defaultTimezone || DEFAULT_TIMEZONE;
  }

  translate = (phrase: string, params: TranslateParams, context: string) => {
    const { translator } = this.settings;

    if (translator) {
      return translator(phrase, params, context);
    }

    let message = phrase;

    if (phrase.includes('||||')) {
      const [singular, plural] = phrase.split('||||');
      const count = params.smartCount || 0;

      message = count === 1 ? singular : plural;
    }

    // Low-level token interpolation
    return message.replace(/%\{(\w+)\}/g, (match, key) => String(params[key] || ''));
  };
}

export default new Core();
