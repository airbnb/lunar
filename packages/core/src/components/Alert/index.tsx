/* eslint-disable prefer-destructuring */

import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import IconFlag from '@airbnb/lunar-icons/lib/interface/IconFlag';
import IconCheckAlt from '@airbnb/lunar-icons/lib/interface/IconCheckAlt';
import IconClose from '@airbnb/lunar-icons/lib/interface/IconClose';
import IconError from '@airbnb/lunar-icons/lib/interface/IconError';
import IconInfo from '@airbnb/lunar-icons/lib/interface/IconInfo';
import IconWarning from '@airbnb/lunar-icons/lib/interface/IconWarning';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import Row from '../Row';
import Spacing from '../Spacing';
import Text from '../Text';
import T from '../Translate';
import IconButton from '../IconButton';
import { STATUSES } from '../../constants';

const alertColorTypePropType = mutuallyExclusiveTrueProps(...STATUSES);

export type Props = {
  /** Content within the label. */
  children?: React.ReactNode;
  /** Dangerous/failure status (red). */
  danger?: boolean;
  /** Whether to hide the corresponding status icon. */
  hideStatusIcon?: boolean;
  /** Informational status (blue). */
  info?: boolean;
  /** Display the alert as inline. */
  inline?: boolean;
  /** Notice status. */
  notice?: boolean;
  /** Successful status (green). */
  success?: boolean;
  /** The title of the alert, bold text at the top of the box. */
  title: NonNullable<React.ReactNode>;
  /** Warning status (yellow). */
  warning?: boolean;
  /** Callback fired when the alert is closed. */
  onClose?: () => void;
};

/** Classify content through the use of colorful alerts. */
export class Alert extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    danger: alertColorTypePropType,
    info: alertColorTypePropType,
    notice: alertColorTypePropType,
    success: alertColorTypePropType,
    warning: alertColorTypePropType,
  };

  static defaultProps = {
    children: null,
    danger: false,
    hideStatusIcon: false,
    info: false,
    inline: false,
    notice: false,
    success: false,
    warning: false,
  };

  render() {
    const {
      cx,
      children,
      danger,
      hideStatusIcon,
      info,
      inline,
      notice,
      styles,
      success,
      theme,
      title,
      warning,
      onClose,
    } = this.props;
    const { color, unit } = theme!;
    let StatusIcon = null;
    let iconColor = color.accent.text;

    if (danger) {
      StatusIcon = IconError;
      iconColor = color.core.danger[3];
    } else if (info) {
      StatusIcon = IconInfo;
      iconColor = color.core.primary[3];
    } else if (notice) {
      StatusIcon = IconFlag;
      iconColor = color.core.secondary[3];
    } else if (success) {
      StatusIcon = IconCheckAlt;
      iconColor = color.core.success[3];
    } else if (warning) {
      StatusIcon = IconWarning;
      iconColor = color.core.warning[3];
    }

    return (
      <div
        className={cx(
          styles.alert,
          inline && styles.alert_inline,
          !hideStatusIcon && !!StatusIcon && styles.alert_statusIcon,
          danger && styles.alert_danger,
          info && styles.alert_info,
          notice && styles.alert_notice,
          success && styles.alert_success,
          warning && styles.alert_warning,
        )}
      >
        <Row
          middleAlign={!children}
          after={
            onClose && (
              <IconButton onClick={onClose}>
                <IconClose
                  accessibilityLabel={T.phrase('Close', {}, 'Close the alert')}
                  size={unit * 3}
                />
              </IconButton>
            )
          }
          before={
            !hideStatusIcon &&
            StatusIcon && <StatusIcon decorative color={iconColor} size={unit * 3} />
          }
        >
          <Text bold>{title}</Text>

          {children && (
            <Spacing top={1}>
              <Text>{children}</Text>
            </Spacing>
          )}
        </Row>
      </div>
    );
  }
}

export default withStyles(
  ({ color, unit, pattern }) => ({
    alert: {
      ...pattern.box,
      display: 'block',
      alignItems: 'start',
      position: 'relative',
      borderColor: color.core.neutral[1],
      backgroundColor: color.accent.bg,
      overflow: 'hidden',
      padding: unit * 3,

      ':before': {
        content: '" "',
        backgroundColor: color.accent.text,
        position: 'absolute',
        top: 0,
        left: 0,
        width: unit / 2,
        height: '100%',
      },
    },

    alert_inline: {
      display: 'inline-block',
    },

    alert_danger: {
      ':before': {
        backgroundColor: color.core.danger[3],
      },
    },

    alert_info: {
      ':before': {
        backgroundColor: color.core.primary[3],
      },
    },

    alert_notice: {
      ':before': {
        backgroundColor: color.core.secondary[3],
      },
    },

    alert_success: {
      ':before': {
        backgroundColor: color.core.success[3],
      },
    },

    alert_warning: {
      ':before': {
        backgroundColor: color.core.warning[3],
      },
    },

    alert_statusIcon: {
      paddingLeft: unit * 2.5,
    },
  }),
  {
    passThemeProp: true,
  },
)(Alert);
