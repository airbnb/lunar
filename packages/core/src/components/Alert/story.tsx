import React from 'react';
import { action } from '@storybook/addon-actions';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '../Text';
import MutedButton from '../MutedButton';
import Spacing from '../Spacing';
import Alert from '.';

export default {
  title: 'Core/Alert',
  parameters: {
    inspectComponents: [Alert],
  },
};

export function defaultAndStatusClassifiedAlerts() {
  return (
    <>
      <Alert title="Alert title" />
      <br />
      <Alert notice title="Notice alert title" />
      <br />
      <Alert info title="Info alert title" />
      <br />
      <Alert success title="Success alert title" />
      <br />
      <Alert warning title="Warning alert title" />
      <br />
      <Alert danger title="Danger alert title" />
    </>
  );
}

defaultAndStatusClassifiedAlerts.story = {
  name: 'Default and status classified alerts.',
};

export function withNoIcons() {
  return (
    <>
      <Alert hideStatusIcon title="Alert title" />
      <br />
      <Alert notice hideStatusIcon title="Notice alert title" />
      <br />
      <Alert info hideStatusIcon title="Info alert title" />
      <br />
      <Alert success hideStatusIcon title="Success alert title" />
      <br />
      <Alert warning hideStatusIcon title="Warning alert title" />
      <br />
      <Alert danger hideStatusIcon title="Danger alert title" />
    </>
  );
}

withNoIcons.story = {
  name: 'With no icons.',
};

export function withACloseButton() {
  return (
    <>
      <div>
        <Alert title="Alert title" onClose={action('onClose')} />

        <br />

        <Alert title="Alert title" onClose={action('onClose')}>
          <Text>Supporting content that is descriptive and helpful would go here</Text>
        </Alert>

        <br />

        <Alert danger title="Danger alert title" onClose={action('onClose')}>
          <Text>Supporting content that is descriptive and helpful would go here</Text>
        </Alert>
      </div>
    </>
  );
}

withACloseButton.story = {
  name: 'With a close button.',
};

export function inlineStory() {
  return (
    <Alert inline notice title="Inline notice alert title">
      <Text>Supporting content that is descriptive and helpful would go here</Text>
    </Alert>
  );
}

inlineStory.story = {
  name: 'Inline.',
};

export function withSupportingContent() {
  return (
    <Alert danger title="Something failed horribly" onClose={action('onClose')}>
      <MutedButton inverted onClick={action('onClick')}>
        Retry the thing
      </MutedButton>
    </Alert>
  );
}

withSupportingContent.story = {
  name: 'With supporting content.',
};

export function withWrappingContent() {
  return (
    <Alert info title="Something failed horribly" onClose={action('onClose')}>
      <Text>
        <LoremIpsum />
      </Text>

      <Spacing top={1}>
        <MutedButton inverted onClick={action('onClick')}>
          Retry the thing
        </MutedButton>
      </Spacing>
    </Alert>
  );
}

withWrappingContent.story = {
  name: 'With wrapping content.',
};
