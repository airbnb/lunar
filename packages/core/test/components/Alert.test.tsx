import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import { unwrapHOCs } from '@airbnb/lunar-test-utils';
import Alert from '../../src/components/Alert';
import Spacing from '../../src/components/Spacing';
import { STATUSES as BASE_STATUSES } from '../../src/constants';

const STATUSES = [...BASE_STATUSES];

function unwrap(element: any): Enzyme.ShallowWrapper<any, any> {
  return unwrapHOCs(shallow(element), 'Alert', {}, { render: true });
}

describe('<Alert />', () => {
  it('renders title', () => {
    const title = 'Foo';
    const wrapper = unwrap(<Alert title={title} />);

    expect(wrapper.contains(title)).toBe(true);
  });

  it('renders close button', () => {
    const wrapper = shallow(<Alert title="Title" onClose={() => {}} />).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inline', () => {
    const wrapper = shallow(<Alert inline title="Title" />).dive();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders space between the title and content', () => {
    const content = 'Foo';
    const wrapper = unwrap(<Alert title="Title">{content}</Alert>);

    expect(wrapper.contains(content)).toBe(true);
    expect(wrapper.find(Spacing)).toHaveLength(1);
    expect(
      unwrapHOCs(wrapper.find(Spacing), 'Spacing', {}, { render: true }).contains(content),
    ).toBe(true);
  });

  describe('renders statuses', () => {
    STATUSES.forEach(status => {
      it(`renders ${status} alert`, () => {
        const wrapper = shallow(<Alert {...{ [status]: true }} title={status} />).dive();

        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
