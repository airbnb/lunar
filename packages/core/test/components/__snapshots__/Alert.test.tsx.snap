// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> renders close button 1`] = `
<div
  className="alert_1ypwfdz"
>
  <Row
    after={
      <withStyles(IconButton)
        onClick={[Function]}
      >
        <IconClose
          accessibilityLabel="Close"
          color="currentColor"
          flip={false}
          flipVertical={false}
          inline={false}
          size={24}
        />
      </withStyles(IconButton)>
    }
    before={null}
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      Title
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders inline 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_inline_36rlri"
>
  <Row
    before={null}
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      Title
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders danger alert 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_statusIcon_l3ubg1-o_O-alert_danger_67ky1g"
>
  <Row
    before={
      <IconError
        color="#DF2323"
        decorative={true}
        flip={false}
        flipVertical={false}
        inline={false}
        size={24}
      />
    }
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      danger
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders info alert 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_statusIcon_l3ubg1-o_O-alert_info_1mewzye"
>
  <Row
    before={
      <IconInfo
        color="#6f44ff"
        decorative={true}
        flip={false}
        flipVertical={false}
        inline={false}
        size={24}
      />
    }
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      info
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders muted alert 1`] = `
<div
  className="alert_1ypwfdz"
>
  <Row
    before={null}
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      muted
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders notice alert 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_statusIcon_l3ubg1-o_O-alert_notice_uvd2g5"
>
  <Row
    before={
      <IconFlag
        color="#1561B2"
        decorative={true}
        flip={false}
        flipVertical={false}
        inline={false}
        size={24}
      />
    }
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      notice
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders success alert 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_statusIcon_l3ubg1-o_O-alert_success_1law0a4"
>
  <Row
    before={
      <IconCheckAlt
        color="#4AA83E"
        decorative={true}
        flip={false}
        flipVertical={false}
        inline={false}
        size={24}
      />
    }
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      success
    </Text>
  </Row>
</div>
`;

exports[`<Alert /> renders statuses renders warning alert 1`] = `
<div
  className="alert_1ypwfdz-o_O-alert_statusIcon_l3ubg1-o_O-alert_warning_1hlm3ds"
>
  <Row
    before={
      <IconWarning
        color="#FFB400"
        decorative={true}
        flip={false}
        flipVertical={false}
        inline={false}
        size={24}
      />
    }
    middleAlign={true}
  >
    <Text
      bold={true}
    >
      warning
    </Text>
  </Row>
</div>
`;
